packages/lesswrong/server/pingbacks.js

```diff
@@ -0,0 +1,52 @@
+import cheerio from 'cheerio';
+import { parsePath } from '../components/linkPreview/HoverPreviewLink.jsx';
+import { parseRoute } from 'meteor/vulcan:core';
+import { hostIsOnsite, getUrlClass } from '../lib/routeUtil';
+import deepmerge from 'deepmerge';
+
+export const htmlToPingbacks = async (html) => {
+  const URLClass = getUrlClass()
+  const links = extractLinks(html);
+  const pingbacks = [];
+  
+  for (let link of links)
+  {
+    // HACK: Parse URLs as though relative to example.com because they have to
+    // be the builtin URL parser needs them to be relative to something with a
+    // domain, and the domain doesn't matter at all except in whether or not
+    // it's in the domain whitelist (which it will only be if it's overridden
+    // by an absolute link).
+    const linkTargetAbsolute = new URLClass(link, 'http://example.com/');
+    
+    if (hostIsOnsite(linkTargetAbsolute.host)) {
+      const onsiteUrl = linkTargetAbsolute.pathname + linkTargetAbsolute.search + linkTargetAbsolute.hash;
+      const parsedUrl = parseRoute({
+        location: parsePath(onsiteUrl),
+        onError: (pathname) => {
+          // Ignore malformed links
+        }
+      });
+      
+      if (parsedUrl?.currentRoute?.getPingback) {
+        const pingback = await parsedUrl.currentRoute.getPingback(parsedUrl);
+        if (pingback)
```

Comment:
This is pretty confusing. getPinpback returns an object of the shape `{ Posts: [postId] }`, but then we push it here into an array, and then later on we deepmerge it. This seems pretty confusing. Why don't we just have it return an id? Or at least explicitly handle the array merging instead of doing a magical deepmerge. 

A lot of this seems to be caused by dealing with blackbox objects, so reiterating wanting all of this to have schemas.

Reply:
Changed the shape to `{collectionName, documentId}` which is more intuitive. (This doesn't support one-link-multiple-pingbacks, but that can be added later if we end up doing it.)